File: scala/say/src/main/scala/Say.scala
object Say {
  def quotRem(x: Long, y: Long) = {
    val div = x / y
    val rem = x % y
    (div, rem)
  }

  val basesValues =
    List(
      "",
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen")

  def bases(n: Long) =
    if (n < 1 || n >= basesValues.length) None
    else Some(basesValues(n.toInt))

  val tensValues =
    List(
      "",
      "",
      "twenty",
      "thirty",
      "forty",
      "fifty",
      "sixty",
      "seventy",
      "eighty",
      "ninety")

  def tens(n: Long) = {
    if (n < 20) bases(n)
    else {
      val (count, remainder) = quotRem(n, 10)
      val countStr = tensValues(count.toInt)
      val baseStr = bases(remainder).fold("") {
        "-" + _
      }
      Some(countStr + baseStr)
    }
  }

  def hundreds(n: Long) = {
    if (n < 100) tens(n)
    else {
      val (count, remainder) = quotRem(n, 100)
      val tensStr = tens(remainder).fold("") {
        " " + _
      }
      bases(count).flatMap { x => Some(x + " hundred" + tensStr) }
    }
  }

  def chunk(str: String)(n: Long) = hundreds(n).flatMap { x => Some(x + " " + str) }

  def thousands = chunk("thousand")_

  def millions = chunk("million")_

  def billions = chunk("billion")_

  def parts(n: Long) = {
    val (billionsCount, billionsRemainder) = quotRem(n, 1000000000L)
    val (millionsCount, millionsRemainder) = quotRem(billionsRemainder, 1000000L)
    val (thousandsCount, remainder) = quotRem(millionsRemainder, 1000L)
    (billionsCount, millionsCount, thousandsCount, remainder)
  }

  def inEnglish(n: Long) = {
    if (n < 0 || n >= 1000000000000L) None
    else if (n == 0) Some("zero")
    else {
      val (billionsCount, millionsCount, thousandsCount, remainder) = parts(n)
      val partStrings = List(billions(billionsCount),
        millions(millionsCount),
        thousands(thousandsCount),
        hundreds(remainder))
      val str = partStrings
        .collect { case Some(x) => x }
        .mkString(" ")
      Some(str)
    }
  }
}

File: nim/say/say.nim
import strutils

type
  Number = object
    billions: int
    millions: int
    thousands: int
    hundreds: int
    remainder: int

proc initNumber(n: int64): Number =
  var remainder = n

  result.billions = int(remainder div 1_000_000_000)
  remainder = remainder mod 1_000_000_000

  result.millions = int(remainder div 1_000_000)
  remainder = remainder mod 1_000_000

  result.thousands = int(remainder div 1_000)
  remainder = remainder mod 1_000

  result.hundreds = int(remainder div 100)
  result.remainder = int(remainder mod 100)

proc sayBelowTwenty(n: int): string =
  case n
    of 0: ""
    of 1: "one"
    of 2: "two"
    of 3: "three"
    of 4: "four"
    of 5: "five"
    of 6: "six"
    of 7: "seven"
    of 8: "eight"
    of 9: "nine"
    of 10: "ten"
    of 11: "eleven"
    of 12: "twelve"
    of 13: "thirteen"
    of 14: "fourteen"
    of 15: "fiveteen"
    of 16: "sixteen"
    of 17: "seventeen"
    of 18: "eighteen"
    of 19: "nineteen"
    else: raise newException(ValueError, "Number not below 20")

proc sayBelowHundred(n: int): string =
  proc say(base: string): string =
    let remainder = n mod 10
    if remainder == 0:
      base
    else:
      base & "-" & sayBelowTwenty(remainder)

  case n
    of 0..19: sayBelowTwenty(n)
    of 20..29: say("twenty")
    of 30..39: say("thirty")
    of 40..49: say("forty")
    of 50..59: say("fifty")
    of 60..69: say("sixty")
    of 70..79: say("seventy")
    of 80..89: say("eighty")
    of 90..99: say("ninety")
    else: raise newException(ValueError, "Number not below 100")

proc sayBelowThousand(n: int): seq[string] =
  let hundreds = n div 100
  let remainder = n mod 100
  
  if hundreds > 0:
    result.add(sayBelowTwenty(hundreds))
    result.add("hundred")

  if remainder > 0:
    result.add(sayBelowHundred(remainder))

proc parts(n: int64): seq[string] =
  let number = initNumber(n)

  if number.billions > 0:
    result.add(sayBelowThousand(number.billions))
    result.add("billion")

  if number.millions > 0:
    result.add(sayBelowThousand(number.millions))
    result.add("million")

  if number.thousands > 0:
    result.add(sayBelowThousand(number.thousands))
    result.add("thousand")

  if number.hundreds > 0:
    result.add(sayBelowHundred(number.hundreds))
    result.add("hundred")

  if number.remainder > 0:
    result.add(sayBelowHundred(number.remainder))

proc say*(n: int64): string =
  if n < 0 or n >= 1_000_000_000_000:
    raise newException(ValueError, "Invalid number")

  if n == 0:
    return "zero"

  parts(n).join(" ");
